src/doc/rustc/src/command-line-arguments.md

```diff
@@ -233,7 +233,12 @@ This flag, when combined with other flags, makes them produce extra output.
 
 This flag allows you to pass the name and location of an external crate that
 will be linked into the crate you are building. This flag may be specified
-multiple times. The format of the value should be `CRATENAME=PATH`.
+multiple times. This flag takes an argument with either of the following
+formats:
+
+* `CRATENAME=PATH` — Indicates the given crate is found at the given path.
+* `CRATENAME` — Indicates the given crate may be found in the search path,
+  such as within the sysroot or via the `-L` flag.
```

Comment:
Do we have tests ensuring that sysroot crates, e.g. private rustc details emit an error when used on a stable compiler? (Also when an explicit path is used to them, not just a search.)

Reply:
What happens if both `--extern CRATENAME=PATH` and `--extern CRATENAME` are specified for the same `CRATENAME`? It would be nice to document that.
AFAIR, multiple `--extern CRATENAME=PATH`s with the same `CRATENAME` are possible if they lead to different kinds of libraries, like `rlib` and `dylib`.

Reply:
There are about 30 tests doing various things with pathless `--extern` (search for `// compile-flags:.*--extern`).

I have added some more specific tests:

- `run-make-fulldeps/extern-flag-pathless`: Shows what happens when you mix pathless with path `--extern` flags. Also demonstrates the preference of rlib over dylib.
- `ui-fulldeps/pathless-extern-unstable.rs`: Shows `--extern rustc` is not allowed.
- `ui/pathless-extern-ok.rs`: Basic test for a sysroot crate.

I don't have a good intuition of how sophisticated rustc tests should be. They are also a bit disorganized, so I didn't know where to stick them, or which style would be preferred. I *think* the dylib tests should be safe from a cross platform standpoint (I believe they only run on Linux?).

I updated the documentation with some more detail. I never really know how detailed rustc docs should be. I intentionally left the prefer-dynamic algorithm vague. It is described somewhat in [`src/librustc_metadata/dependency_format.rs`](https://github.com/rust-lang/rust/blob/master/src/librustc_metadata/dependency_format.rs). I figure since it is complex, and possibly subject to change, it may not be worthwhile. I can add more or remove some if desired.